diff -ru pthreads-w32/pthreads.2/Makefile pthreads-w32-mod/pthreads.2/Makefile
--- pthreads-w32/pthreads.2/Makefile	2018-08-08 09:44:21.000000000 +0800
+++ pthreads-w32-mod/pthreads.2/Makefile	2018-08-08 11:12:15.000000000 +0800
@@ -13,7 +13,7 @@
 
 DEVROOT	= C:\pthreads
 
-DLLDEST	= $(DEVROOT)\dll
+DLLDEST	= $(DEVROOT)\bin
 LIBDEST	= $(DEVROOT)\lib
 HDRDEST	= $(DEVROOT)\include
 
@@ -469,6 +469,9 @@
 
 
 install:
+	if not exist $(DLLDEST) mkdir $(DLLDEST)
+	if not exist $(LIBDEST) mkdir $(LIBDEST)
+	if not exist $(HDRDEST) mkdir $(HDRDEST)
 	copy pthread*.dll $(DLLDEST)
 	copy pthread*.lib $(LIBDEST)
 	copy pthread.h $(HDRDEST)
diff -ru pthreads-w32/pthreads.2/pthread.h pthreads-w32-mod/pthreads.2/pthread.h
--- pthreads-w32/pthreads.2/pthread.h	2018-08-08 09:44:21.000000000 +0800
+++ pthreads-w32-mod/pthreads.2/pthread.h	2018-08-08 11:10:37.000000000 +0800
@@ -313,7 +313,7 @@
 #endif
 #endif
 
-#if !defined(HAVE_STRUCT_TIMESPEC)
+#if !defined(HAVE_STRUCT_TIMESPEC) && (_MSC_VER < 1900)
 #define HAVE_STRUCT_TIMESPEC
 #if !defined(_TIMESPEC_DEFINED)
 #define _TIMESPEC_DEFINED
